sdk/python/kfp/dsl/legacy_data_passing_adaptor.py

```diff
@@ -0,0 +1,144 @@
+# Copyright 2021 The Kubeflow Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Adaptors to support legacy data passing.
+
+The adaptor offers backward compatiblity for pipelines utilize legacy data
+passing. Passing parameters to artifacts or vice versa could result unexpected
+runtime errors. It is highly recommended to not rely on the adaptors. We may
+remove the adaptors in the later releases.
+"""
+from typing import Type, Union
+
+from kfp import dsl
+from kfp.components import _naming
+from kfp.dsl import dsl_utils
+from kfp.dsl import io_types
+from kfp.dsl import type_utils
+from kfp.pipeline_spec import pipeline_spec_pb2 as pb
+
+_PARAMETER_TYPES = Union[str, int, float, bool, dict, list]
+
+# Global index to generate unique and stable adaptor name.
+adaptor_op_index = 0
+
+
+class BaseAdaptor:
+  """Base class of legacy data passing adaptor.
+
+  Attributes:
+    INPUT_KEY: The name of the component input.
+    OUTPUT_KEY: The name of the component output.
+    task_spec: The IR PipelineTaskSpec of the component.
+    component_spec: The IR ComponentSpec of the component.
+    container_spec: The IR PipelineContainerSpec of the component.
+  """
+  INPUT_KEY = 'input'
+  OUTPUT_KEY = 'output'
+  _IMAGE = 'alpine:3.14'
+
+  def __init__(self, base_name: str):
+    """Inits partially the task_spec.
+
+    The task input spec will be assigned during the components connecting phase
+    (in `kfp.dsl._component_bridage`).
+
+    Args:
+      base_name: The base name of the adaptor op. Is will be suffixed with a
```

Comment:
Just a thought:
When upstream is a task output, the adapter id suffix can be `<producerTaskId>_<output_name>`
When upstream is a graph input, the adapter id suffix can be `graph_input_<output_name>`
When upstream is constant value string, the adapter id suffix can be `<consumerTaskId>_<input_name>`

Reply:
Agree that would be nicer, but don't think its worth the complexity to optimize adaptor as we still perceive it as an temporary solution.